diff -ur camping-1.4.2-o/bin/camping camping-1.4.2/bin/camping
--- camping-1.4.2-o/bin/camping	2006-05-16 16:11:19.000000000 -0600
+++ camping-1.4.2/bin/camping	2006-08-08 15:24:57.000000000 -0600
@@ -10,7 +10,6 @@
 require 'delegate'
 require 'optparse'
 require 'stringio'
-require 'rubygems'
 require 'camping'
 
 host = '0.0.0.0'
diff -ur camping-1.4.2-o/examples/blog/blog.rb camping-1.4.2/examples/blog/blog.rb
--- camping-1.4.2-o/examples/blog/blog.rb	2006-05-11 12:53:23.000000000 -0600
+++ camping-1.4.2/examples/blog/blog.rb	2006-08-08 15:26:40.000000000 -0600
@@ -1,8 +1,6 @@
 #!/usr/bin/env ruby
 
-$:.unshift File.dirname(__FILE__) + "/../../lib"
-require 'rubygems'
-require_gem 'camping', '>=1.4'
+require 'camping'
 require 'camping/session'
   
 Camping.goes :Blog
diff -ur camping-1.4.2-o/examples/campsh/campsh.rb camping-1.4.2/examples/campsh/campsh.rb
--- camping-1.4.2-o/examples/campsh/campsh.rb	2006-05-03 14:12:18.000000000 -0600
+++ camping-1.4.2/examples/campsh/campsh.rb	2006-08-08 15:27:21.000000000 -0600
@@ -1,7 +1,6 @@
 #!/usr/bin/env ruby
 
-$:.unshift File.dirname(__FILE__) + "/../../lib"
-%w(rubygems redcloth camping acts_as_versioned).each { |lib| require lib }
+%w(redcloth camping acts_as_versioned).each { |lib| require lib }
   
 Camping.goes :CampSh
 
diff -ur camping-1.4.2-o/examples/charts/charts.rb camping-1.4.2/examples/charts/charts.rb
--- camping-1.4.2-o/examples/charts/charts.rb	2006-05-03 14:12:18.000000000 -0600
+++ camping-1.4.2/examples/charts/charts.rb	2006-08-08 15:27:14.000000000 -0600
@@ -1,8 +1,6 @@
 #!/usr/bin/env ruby
 
-$:.unshift File.dirname(__FILE__) + "/../../lib"
 $:.unshift File.dirname(__FILE__)
-require 'rubygems'
 require 'fileutils'
 require 'camping'
 require 'rvg/rvg'
diff -ur camping-1.4.2-o/examples/tepee/tepee.rb camping-1.4.2/examples/tepee/tepee.rb
--- camping-1.4.2-o/examples/tepee/tepee.rb	2006-05-03 14:12:18.000000000 -0600
+++ camping-1.4.2/examples/tepee/tepee.rb	2006-08-08 15:26:53.000000000 -0600
@@ -1,6 +1,5 @@
 #!/usr/bin/ruby
-$:.unshift File.dirname(__FILE__) + "/../../lib"
-%w(rubygems redcloth camping acts_as_versioned).each { |lib| require lib }
+%w(redcloth camping acts_as_versioned).each { |lib| require lib }
 
 Camping.goes :Tepee
 
diff -ur camping-1.4.2-o/lib/camping-unabridged.rb camping-1.4.2/lib/camping-unabridged.rb
--- camping-1.4.2-o/lib/camping-unabridged.rb	2006-05-16 12:00:12.000000000 -0600
+++ camping-1.4.2/lib/camping-unabridged.rb	2006-08-08 15:25:34.000000000 -0600
@@ -28,7 +28,7 @@
 # http://rubyforge.org/projects/mongrel  Mongrel comes with examples
 # in its <tt>examples/camping</tt> directory. 
 #
-%w[rubygems active_record markaby metaid tempfile uri].each { |lib| require lib }
+%w[active_record markaby metaid tempfile uri].each { |lib| require lib }
 
 # == Camping 
 #
diff -ur camping-1.4.2-o/lib/camping.rb camping-1.4.2/lib/camping.rb
--- camping-1.4.2-o/lib/camping.rb	2006-05-16 12:00:21.000000000 -0600
+++ camping-1.4.2/lib/camping.rb	2006-08-08 15:25:10.000000000 -0600
@@ -1,4 +1,4 @@
-%w[rubygems active_record markaby metaid tempfile uri].each{|l|require l}
+%w[active_record markaby metaid tempfile uri].each{|l|require l}
 module Camping;C=self;F=__FILE__;S=IO.read(F).gsub(/_+FILE_+/,F.dump)
 module Helpers;def R c,*args;p=/\(.+?\)/;args.inject(c.urls.find{|x|x.scan(p).
 size==args.size}.dup){|str,a|str.sub(p,(a.__send__(a.class.primary_key)rescue
